[GDB] Fix pretty-printer crash for sycl::item MOffset

ianayl opened the pull request:
The pretty printer looks for an `MOffset` field in `sycl::item` impl without considering whether or not offset has been enabled (Offset was deprecated in SYCL 2020). Add a check to avoid printing `MOffset` if offset has been disabled.

uditagarwal97 commented:
Do you know if it's feasible to add a test for this change? I don't recall seeing tests for `sycl/gdb/libsycl.so-gdb.py` TBH, so not sure if it's feasible.

ianayl commented:
> Do you know if it's feasible to add a test for this change? I don't recall seeing tests for sycl/gdb/libsycl.so-gdb.py TBH, so not sure if it's feasible.

Looks like we don't have any testing for `libsycl.so-gdb.py`; the closest thing we have is https://github.com/intel/llvm/blob/sycl/sycl/test/gdb/printers.cpp, but this does not actually test the python printer script itself, but rather checks AST layouts. 

Should I still go about updating said file to contain `sycl::item`? I had initially tried to extend this file to check that our classes contain fields that are as expected by the pretty printer, but as I actually started writing the test, I realized that since we never actually test-run `libsycl.so-gdb.py`, the test seems rather useless for my sakes: I assume we are trying to explicitly check if the pretty-printer has bugs in it, not whether or not SYCL RT has changes to it instead.

uditagarwal97 commented:
> Should I still go about updating said file to contain `sycl::item`? 

Yes

ianayl commented:
I added a check in `sycl/test/gdb/printers.cpp` for `sycl::item`, checking for both `MOffset` and no `MOffset`: But as I mentioned prior, this test itself never actually runs the gdb printer 😅 it is effectively a lit test that checks SYCL constructs are as we expect them to be. I've added the test as requested, but I don't know if it does anything useful...

uditagarwal97 commented:
> I added a check in `sycl/test/gdb/printers.cpp` for `sycl::item`, checking for both `MOffset` and no `MOffset`: But as I mentioned prior, this test itself never actually runs the gdb printer 😅 it is effectively a lit test that checks SYCL constructs are as we expect them to be. I've added the test as requested, but I don't know if it does anything useful...

Thanks. I think the test is helpful. The test `sycl/test/gdb/printers.cpp` checks the layout of SYCL objects and what `libsycl.so-gdb.py` assumes. If in future, the layout of `sycl::item` changes further, we should be able to catch regressions like this.

ianayl commented:
@intel/llvm-gatekeepers PR should be good to merge: CI error is unrelated as this PR makes no modifications to DPC++ itself.

ianayl commented:
> If the test failure is unrelated, can you share the link to the Github issue tracking this failure or create a new GH issue for it?

That's a good point

Issue tracking both `USM/memops2d/copy2d_device_to_host.cpp` and `copy2d_host_to_device.cpp` failures: https://github.com/intel/llvm/issues/19749

`copy2d_host_to_device.cpp` is failing here
